src/mpid/ch4/shm/posix/eager/fbox/fbox_recv.h

```diff
@@ -40,19 +40,41 @@ MPIDI_POSIX_eager_recv_begin(MPIDI_POSIX_eager_recv_transaction_t * transaction)
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_POSIX_FBOX_EAGER_RECV_BEGIN);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_POSIX_FBOX_EAGER_RECV_BEGIN);
 
-    int j, local_rank, grank;
+    int j = 0, local_rank, grank;
     MPIDI_POSIX_fastbox_t *fbox_in;
     int mpi_errno = MPIDI_POSIX_NOK;
 
     /* Rather than polling all of the fastboxes on every loop, do a small number and rely on calling
      * this function again if needed. */
-    for (j = 0; j < MPIR_CVAR_CH4_POSIX_EAGER_FBOX_BATCH_SIZE; j++) {
+    for (j = 0; j < MPIR_CVAR_CH4_POSIX_EAGER_FBOX_POLL_CACHE_SIZE +
+         MPIR_CVAR_CH4_POSIX_EAGER_FBOX_BATCH_SIZE; j++) {
+
+        /* Before polling *all* of the fastboxes, poll the ones that are most likely to have messages
+         * (where receives have been preposted). */
+        if (j < MPIR_CVAR_CH4_POSIX_EAGER_FBOX_POLL_CACHE_SIZE) {
+            /* Get the next fastbox to poll. */
+            local_rank = MPIDI_POSIX_eager_fbox_control_global.first_poll_local_ranks[j];
+        }
+        /* If we have finished the cached fastboxes, continue polling the rest of the fastboxes
+         * where we left off last time. */
+        else {
+            /* Figure out the next fastbox to poll by incrementing the counter. */
+            MPIDI_POSIX_eager_fbox_control_global.first_poll_local_ranks
+                [MPIR_CVAR_CH4_POSIX_EAGER_FBOX_POLL_CACHE_SIZE] =
+                (MPIDI_POSIX_eager_fbox_control_global.first_poll_local_ranks
+                 [MPIR_CVAR_CH4_POSIX_EAGER_FBOX_POLL_CACHE_SIZE] +
+                 1) % (int16_t) MPIDI_POSIX_global.num_local;
+            local_rank =
+                MPIDI_POSIX_eager_fbox_control_global.first_poll_local_ranks
+                [MPIR_CVAR_CH4_POSIX_EAGER_FBOX_POLL_CACHE_SIZE];
+        }
```

Comment:
The code in the `else` branch is a little hard to read. Can we reformat it with a local variable instead of using the long `MPIDI_POSIX_eager_fbox_control_global.first_poll_local_ranks[MPIR_CVAR_CH4_POSIX_EAGER_FBOX_POLL_CACHE_SIZE]` every time ?

Reply:
Makes sense. 👍